keps/sig-storage/20200609-csi-service-account-token.md

```diff
@@ -0,0 +1,129 @@
+2.  CSI driver will have access to the secrets of pods that do not use it
+    because the CSI driver should have a `hostPath` volume for the `pods`
+    subdirectory to read the token.
+3.  The audience of the token is defaulted to kube apiserver.
+4.  The token is not guaranteed to be available (e.g.
+    `automountServiceAccountToken=false`).
+
+### User stories
+
+- HashiCorp Vault provider for secret store CSI driver requires service
+  account token of the pods they are mounting secrets at to authenticate to
+  Vaults. The provisioned secrets also have given TTL in Vault, so it is
+  neccessary get tokens after the inital mount.
+- Cert manager CSI dirver will create CertificateRequests on behalf of the
+  pods.
+- Amazon EFS CSI driver wants the service account tokens of pods to exchange for
+  AWS credentials.
+
+## Proposal
+
+Kubelet should inject audience-bounded service account tokens requested by the
+CSI driver from the `TokenRequest` API in `VolumeContext` to
+`NodePublishVolume`.
+
+### CSIDriver Object
+
+```go
+// CSIDriverSpec is the specification of a CSIDriver.
+type CSIDriverSpec struct {
+    ... // existing fields
+
+    RequiresRemount *bool
+    ServiceAccountTokenAudience *string
+}
+```
+
+`ServiceAccountTokenAudience` will be used to set on the created token, which
+will be validated to not be empty string (i.e. kube apiserver).
+`RequiresRemount` is optional, which should be only set when the mounted volumes
+by the CRI driver have TTL and require re-validation on the token.
+
+The token will be set in `VolumeContext`:
+
+```go
+"csi.storage.k8s.io/serviceAccount.token": token
+```
+
+With the new fields:
+
+1.  The issued token for CSI driver cannot authenticated to kube apiserver.
+    Storage provider is expected to call TokenReview api where audience is
+    specified to validate the token.
+2.  If storage provider requires periodic authentication, `RequiresRemount` can
+    be used to trigger a sequence of `NodePublishVolume` calls. Each call will
+    have valid token in the `VolumeContext`.
+
+### Example Workflow
+
+Take the Vault provider for secret store CSI driver as an example:
+
+1.  Create `CSIDriver` object with `ServiceAccountTokenAudience=vault` and
+    `RequiresRemount=true`
+2.  When the volume manager of kubelet sees a new volume, the pod object in
+    `mountedPods` will have `requiresRemound=true` after `MarkRemountRequired`
+    is called. `MarkRemountRequired` will call into `RequiresRemount` of the
+    in-tree csi plugin to fetch the `CSIDriver` object.
+3.  Before `NodePublishVolume` call, kubelet will request token from
+    `TokenRequest` api with `audience=vault`.
+4.  The token will be specifed in `VolumeContext` to `NodePublishVolume` call.
+5.  Every 0.1 second, the reconciler component of volume manager will remount
+    the volume in case the vault secrets expire and re-login is required.
```

Comment:
> What if one of these remounts (NodePublish?) fails? 
 
return errors? do you have any suggestion?

> What if kubelet is not able to refresh a token after expiration?

the token will be refreshed regularly before expiration.

Does this mean that kubelet must cache "vault tokens" of all running pods and their lifetime, to be able to refresh a token when it expires?

the token is already being cached and refreshed before expiration.

Reply:
Returning errors only produces events to users. A pod keeps running, possibly with expired / stale data, without really knowing about it. Is it something that's acceptable? Or are there cases when a pod that uses a secret must be killed when it looses access to a secret? In that case you should add some note here + code to kill pods.

Reply:
looking at the code, the pod will not be killed if the first volume mount fails. i expect the remounts will adopt the same behavior. @seh what will happen if the mounted secrets expire? will the secrets be deleted so that the users can be notified?

Reply:
The way I wrote my CSI driver, if we can't refresh the service account token, the Vault token, or the projected secret artifacts (from the K/Vv2 or PKI secrets engines), it posts events to the pod, but it doesn't do anything else to the pod. It leaves the old files in place until it can successfully replace them. It does retry these operations, for what it's worth.

Reply:
are the secrets still accessible and valid to use after expiry? 
if so, it is not interesting to remount the volume.
if not, we can leave pods running and hopefully remount will success.

Reply:
Yes, they're still accessible. The driver had already written the content to files that it leaves in place. Only upon a successful refreshing of the secret data does it overwrite the files.

Reply:
but the secrets should not be valid to use right if it has TTL? e.g. the dynamic secrets https://www.vaultproject.io/docs/concepts/lease. and that's the use case for `RemountRequires=true`

then i think it is safe to keep pods running if remount fails because the secrets cannot be used to authenticate.

Reply:
For secrets from the Vault KV engines, the secrets are _probably_ not valid, but they may still be. The TTLs are advisory.

These are trades. In some cases, the container isn't using these secrets continuously, so you have to make the call as to whether it's better to keep the container running with possibly stale secrets (which might be fine for a while) or to kill the container, hoping to eventually be able to refresh the secrets and start the container running again later.

This seems like the relationship between liveness and readiness probes. The latter acknowledges partial degradation, but says that it's not worth killing and starting a replacement container, whether because the start-up cost is high, or because the problem is likely intermittent, or because the container can still do useful work even if it's not ready to serve.

Most containers need these secrets at minimum when first starting. Whether they need them continuously varies. In my CSI driver, it honored a volume parameter specifying whether to try refreshing the secrets, defaulting to not trying. I can imagine evolving that parameter to express that we need the secrets _initially_ (not bothering to refresh them later), _continually_ (killing the container if they ever get stale), or on _a best-effort basis_ (leaving the container running even when we believe the secrets to have gone stale).

Reply:
> killing the container if they ever get stale
if the secret can ever get stale, the secrets provider (e.g. vault) should revoke those secrets, rather than the secrets consumers (e.g. k8s). we can not rely on the credential consumers not to use expired secrets. e.g. the pod cache the secret in memory even if the secret is replaced with another fresh one.

Reply:
Those are fine ideas, but that's not how all the secret engines in Vault work.

Consider an administrator who stores an API key in a Vault K/V secret. Perhaps that administrator usually rotates the key about every other month, leaving a couple of weeks of overlap before revoking the previous key. She might set an advisory TTL on the secret read from Vault for one week, advising consuming applications that they should probably check for a fresh key every week so. An old key might still work for a long while after that week.

Reply:
going back to the original question on what to do when the remount, i would still like to keep the current behavior where the error will only be logged because:
1. stale secrets might still work and it is not security critical for the consumer of credentials (k8s) to stop using the stale secrets.
2. the repeated remount will overcome transient errors.
3. there is no guarantee that killing the container/pod will solve fatal errors.

Reply:
A lot of good discussion here. Can we capture all this somewhere in the KEP?

Reply:
done.